hubble/cmd/common/conn/conn.go

```diff
@@ -30,13 +32,68 @@ var GRPCOptionFuncs []GRPCOptionFunc
 func init() {
 	GRPCOptionFuncs = append(
 		GRPCOptionFuncs,
-		grpcInterceptors,
+		grpcUnaryInterceptors,
+		grpcStreamInterceptors,
 		grpcOptionTLS,
 	)
 }
 
-func grpcInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
-	return grpc.WithUnaryInterceptor(timeout.UnaryClientInterceptor(vp.GetDuration(config.KeyRequestTimeout))), nil
+func grpcUnaryInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
+	option := grpc.WithChainUnaryInterceptor(
+		timeout.UnaryClientInterceptor(vp.GetDuration(config.KeyRequestTimeout)),
+		onReceiveHeaderUnaryInterceptor(logger.Logger, logVersionMismatch()),
+	)
+	return option, nil
+}
+
+func grpcStreamInterceptors(vp *viper.Viper) (grpc.DialOption, error) {
+	option := grpc.WithChainStreamInterceptor(
+		onReceiveHeaderStreamInterceptor(logger.Logger, logVersionMismatch()),
+	)
+	return option, nil
+}
+
+type onReceiveHeader func(log *slog.Logger, header metadata.MD)
+
+// onReceiveHeaderUnaryInterceptor is a gRPC client unary interceptor that retrieves the header
+// metadata and execute the provided function.
+func onReceiveHeaderUnaryInterceptor(log *slog.Logger, fn onReceiveHeader) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req any, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		var header metadata.MD
+		opts = append(opts, grpc.Header(&header))
+		if err := invoker(ctx, method, req, reply, cc, opts...); err != nil {
+			return err
+		}
+		fn(log, header)
+		return nil
+	}
+}
+
+// onReceiveHeaderStreamInterceptor is a gRPC client stream interceptor that retrieves the header
+// metadata and execute the provided function.
+func onReceiveHeaderStreamInterceptor(log *slog.Logger, fn onReceiveHeader) grpc.StreamClientInterceptor {
+	return func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		stream, err := streamer(ctx, desc, cc, method, opts...)
+		if err != nil {
+			return nil, err
+		}
+		// stream.Header() blocks until metadata is ready to read.
+		// This could be:
+		//  - Forever if no metadata is ever sent.
+		//  - After the first read from the stream.
+		//  - After an explicit call to SendHeader() server-side.
+		// To avoid a possible deadlock, perform header extraction in a goroutine tied to
+		// the lifetime of the stream (stream.Header() returns on stream close).
```

Comment:
:+1: I think it's fine especially in the context of the CLI which will open a limited amount of streaming connections.